File: frontend/src/app/core/model/source-summary.ts
/** Whatever `type` a source declares. A new connector is a YAML block, not an enum. */
export type SourceKind = string;

/** Which of the two configuration layers this source's definition came from. */
export type ConfigLayer = 'default' | 'config-dir';

/**
 * One row on the sources screen. `announced` versus `extracted` is the only
 * check nothing else can make: a selector that stops matching loses offers, and
 * fewer offers looks exactly like a quiet day on the market.
 */
export interface SourceSummary {
    readonly id: string;
    readonly kind: SourceKind;
    readonly enabled: boolean;
    readonly lastRunAt: string | null;
    readonly documents: number;
    readonly extracted: number;
    readonly announced: number | null;
    readonly survived: number;
}

/**
 * What `/api/v1/sources` answers: the file that defines the sources, and the sources.
 *
 * <p>The layer is on the envelope and no longer on the row. It is one probe for the whole
 * file — the two configuration layers override each other file by file and never key by key —
 * so a badge per row was asserting something that cannot differ between two rows. Stated once
 * above the table it is simply true, and the table lost a column, which is the only structural
 * relief a seven-column table has on a phone.
 *
 * The file's *path* is deliberately not here: it is a personal datum and this payload is behind
 * every screenshot of the screen. It lives in the panel one deliberate click opens.
 */
export interface SourcesView {
  readonly file: string;
  readonly layer: ConfigLayer;
  readonly sources: readonly SourceSummary[];
}

File: backend/src/main/java/de/codeministry/leadgen/config/SourceSummary.java
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Copyright 2026 Marcello Muscara (codeministry)
 *
 * Licensed under the Apache License, Version 2.0. You may obtain a copy of the
 * License at http://www.apache.org/licenses/LICENSE-2.0
 */
package de.codeministry.leadgen.config;

import java.time.Instant;

/**
 * One row on the sources screen.
 *
 * <p><b>It no longer carries which layer defined it.</b> That value is computed once for the
 * whole file — the two layers override each other file by file and never key by key — and was
 * then stamped onto every row, where it asserted a per-row fact that cannot vary. It sits on
 * {@link SourcesView} now, which is the scope it is true at.
 *
 * @param announced what the documents of the last run stated about themselves, or null
 *                  when none of them says. A mismatch with {@code extracted} means the selectors have
 *                  drifted, which is the one failure that otherwise looks exactly like a quiet market.
 * @param survived  how many of this source's offers cleared the hard filter. Counted over
 *                  the whole archive rather than the last run, because that is what the number is worth
 *                  knowing for.
 */
public record SourceSummary(
        String id,
        String kind,
        boolean enabled,
        Instant lastRunAt,
        int documents,
        int extracted,
        Integer announced,
        int survived) {}
